File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/RtpTransceiverProxy.kt
package com.cloudwebrtc.webrtc.proxy

import com.cloudwebrtc.webrtc.model.RtpTransceiverDirection
import org.webrtc.RtpTransceiver

/** Wrapper around an [RtpTransceiver]. */
class RtpTransceiverProxy(obj: RtpTransceiver) : Proxy<RtpTransceiver>(obj) {
  /** [RtpSenderProxy] of this [RtpTransceiverProxy]. */
  private lateinit var sender: RtpSenderProxy

  /** [RtpReceiverProxy] of this [RtpTransceiverProxy]. */
  private lateinit var receiver: RtpReceiverProxy

  init {
    syncSender()
    syncReceiver()
    addOnSyncListener {
      syncSender()
      syncReceiver()
    }
  }

  /** @return [RtpSenderProxy] of this [RtpTransceiverProxy]. */
  fun getSender(): RtpSenderProxy {
    return sender
  }

  /** @return [RtpReceiverProxy] of this [RtpTransceiverProxy]. */
  fun getReceiver(): RtpReceiverProxy {
    return receiver
  }

  /** Sets [RtpTransceiverDirection] of the underlying [RtpTransceiver]. */
  fun setDirection(direction: RtpTransceiverDirection) {
    obj.direction = direction.intoWebRtc()
  }

  /** Sets receive of the underlying [RtpTransceiver]. */
  fun setRecv(recv: Boolean) {
    var currentDirection = RtpTransceiverDirection.fromWebRtc(obj)
    var newDirection =
        if (recv) {
          when (currentDirection) {
            RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.RECV_ONLY
            RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.RECV_ONLY
            RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_RECV
            RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_RECV
            else -> {
              RtpTransceiverDirection.STOPPED
            }
          }
        } else {
          when (currentDirection) {
            RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.INACTIVE
            RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.INACTIVE
            RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_ONLY
            RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_ONLY
            else -> {
              RtpTransceiverDirection.STOPPED
            }
          }
        }
    if (newDirection != RtpTransceiverDirection.STOPPED) {
      setDirection(newDirection)
    }
  }

  /** Sets send of the underlying [RtpTransceiver]. */
  fun setSend(send: Boolean) {
    var currentDirection = RtpTransceiverDirection.fromWebRtc(obj)
    var newDirection =
        if (send) {
          when (currentDirection) {
            RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.SEND_ONLY
            RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.SEND_ONLY
            RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.SEND_RECV
            RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.SEND_RECV
            else -> {
              RtpTransceiverDirection.STOPPED
            }
          }
        } else {
          when (currentDirection) {
            RtpTransceiverDirection.INACTIVE -> RtpTransceiverDirection.INACTIVE
            RtpTransceiverDirection.SEND_ONLY -> RtpTransceiverDirection.INACTIVE
            RtpTransceiverDirection.SEND_RECV -> RtpTransceiverDirection.RECV_ONLY
            RtpTransceiverDirection.RECV_ONLY -> RtpTransceiverDirection.RECV_ONLY
            else -> {
              RtpTransceiverDirection.STOPPED
            }
          }
        }
    if (newDirection != RtpTransceiverDirection.STOPPED) {
      setDirection(newDirection)
    }
  }

  /** @return mID of the underlying [RtpTransceiver]. */
  fun getMid(): String? {
    return obj.mid
  }

  /** @return Preferred [RtpTransceiverDirection] of the underlying [RtpTransceiver]. */
  fun getDirection(): RtpTransceiverDirection {
    return RtpTransceiverDirection.fromWebRtc(obj)
  }

  /** Stops the underlying [RtpTransceiver]. */
  fun stop() {
    receiver.notifyRemoved()
    obj.stop()
  }

  /**
   * Synchronizes the [RtpSenderProxy] of this [RtpTransceiverProxy] with the underlying
   * [RtpTransceiver].
   */
  private fun syncSender() {
    val newSender = obj.sender
    if (this::sender.isInitialized) {
      sender.replace(newSender)
    } else {
      sender = RtpSenderProxy(newSender)
    }
  }

  /**
   * Synchronizes the [RtpReceiverProxy] of this [RtpTransceiverProxy] with the underlying
   * [RtpTransceiver].
   */
  private fun syncReceiver() {
    val newReceiver = obj.receiver
    if (this::receiver.isInitialized) {
      receiver.replace(newReceiver)
    } else {
      receiver = RtpReceiverProxy(newReceiver)
    }
  }
}

File: ios/Classes/proxy/RtpTransceiverProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpTransceiver`, powering it with additional API.
class RtpTransceiverProxy {
  /// Actual underlying `RTCRtpTransceiver`.
  private var transceiver: RTCRtpTransceiver

  /// `RtpSenderProxy` of this `RtpTransceiverProxy`.
  private var sender: RtpSenderProxy

  /// `RtpReceiverProxy` of this `RtpTransceiverProxy`.
  private var receiver: RtpReceiverProxy

  /// Initializes a new `RtpTransceiverProxy` for the provided
  /// `RTCRtpTransceiver`.
  init(transceiver: RTCRtpTransceiver) {
    self.sender = RtpSenderProxy(sender: transceiver.sender)
    self.receiver = RtpReceiverProxy(receiver: transceiver.receiver)
    self.transceiver = transceiver
  }

  /// Returns `RtpSenderProxy` of this `RtpTransceiverProxy`.
  func getSender() -> RtpSenderProxy {
    self.sender
  }

  /// Returns `RtpReceiverProxy` of this `RtpTransceiverProxy`.
  func getReceiver() -> RtpReceiverProxy {
    self.receiver
  }

  /// Sets the provided `TransceiverDirection` to the underlying
  /// `RTCRtpTransceiver`.
  func setDirection(direction: TransceiverDirection) {
    self.transceiver.setDirection(direction.intoWebRtc(), error: nil)
  }

  /// Changes the preferred `RTCRtpTransceiver` codecs to the provided list of
  /// `RTCRtpCodecCapability`s.
  func setCodecPreferences(capability: [RTCRtpCodecCapability]) {
    self.transceiver.setCodecPreferences(capability)
  }

  /// Sets `recv` direction of the underlying `RTCRtpTransceiver`.
  func setRecv(recv: Bool) {
    let direction = self.getDirection()
    var newDirection = TransceiverDirection.stopped
    if recv {
      switch direction {
      case .inactive:
        newDirection = TransceiverDirection.recvOnly
      case .recvOnly:
        newDirection = TransceiverDirection.recvOnly
      case .sendRecv:
        newDirection = TransceiverDirection.sendRecv
      case .sendOnly:
        newDirection = TransceiverDirection.sendRecv
      case .stopped:
        newDirection = TransceiverDirection.stopped
      }
    } else {
      switch direction {
      case .inactive:
        newDirection = TransceiverDirection.inactive
      case .recvOnly:
        newDirection = .inactive
      case .sendRecv:
        newDirection = .sendOnly
      case .sendOnly:
        newDirection = .sendOnly
      case .stopped:
        newDirection = .stopped
      }
    }

    if newDirection != .stopped {
      self.setDirection(direction: newDirection)
    }
  }

  /// Sets `send` direction of the underlying `RTCRtpTransceiver`.
  func setSend(send: Bool) {
    let direction = self.getDirection()
    var newDirection = TransceiverDirection.stopped
    if send {
      switch direction {
      case .inactive:
        newDirection = .sendOnly
      case .sendOnly:
        newDirection = .sendOnly
      case .sendRecv:
        newDirection = .sendRecv
      case .recvOnly:
        newDirection = .sendRecv
      case .stopped:
        newDirection = .stopped
      }
    } else {
      switch direction {
      case .inactive:
        newDirection = .inactive
      case .sendOnly:
        newDirection = .inactive
      case .sendRecv:
        newDirection = .recvOnly
      case .recvOnly:
        newDirection = .recvOnly
      case .stopped:
        newDirection = .stopped
      }
    }

    if newDirection != .stopped {
      self.setDirection(direction: newDirection)
    }
  }

  /// Returns mID of the underlying `RTCRtpTransceiver`.
  func getMid() -> String? {
    if self.transceiver.mid != nil, self.transceiver.mid.count == 0 {
      return nil
    }
    return self.transceiver.mid
  }

  /// Returns the preferred `RtpTransceiverDirection` of the underlying
  /// `RTCRtpTransceiver`.
  func getDirection() -> TransceiverDirection {
    if self.transceiver.isStopped {
      return TransceiverDirection.stopped
    } else {
      return TransceiverDirection
        .fromWebRtc(direction: self.transceiver.direction)
    }
  }

  /// Stops the underlying `RTCRtpTransceiver`.
  func stop() {
    self.transceiver.stopInternal()
    self.receiver.notifyRemoved()
  }
}
